### PR TITLE
PT-1219: feat: item-ownership-consistency job

### PR DESCRIPTION
**Pull request by augur**

### CODE REVIEW

File: nft/listener/src/main/kotlin/com/rarible/protocol/nft/listener/service/item/ItemOwnershipConsistencyService.kt [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        var value = EthUInt256.ZERO
+        var continuation: OwnershipContinuation? = null
+        do {
+            val filter = OwnershipFilterByItem(
+                contract = itemId.token,
+                tokenId = itemId.tokenId.value,
+                sort = OwnershipFilter.Sort.LAST_UPDATE
+            )
+            val ownerships = ownershipRepository.search(filter.toCriteria(continuation = continuation, limit = limit))
+            continuation = if (ownerships.isEmpty() || ownerships.size < limit) null else ownerships.last()
```
Comment (yakischik) [inline nft/listener/src/main/kotlin/com/rarible/protocol/nft/listener/service/item/ItemOwnershipConsistencyService.kt:71]: we can keep just < limit

File: nft/listener/src/main/kotlin/com/rarible/protocol/nft/listener/service/item/ItemOwnershipConsistencyService.kt [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        itemReduceService.update(item.token, item.tokenId).awaitFirstOrNull()
+        logger.info("Attempt finished for item ${item.id}")
+        return itemRepository.findById(item.id).awaitSingle()
+    }
+
+    private suspend fun getOwnershipsTotalSupply(itemId: ItemId, limit: Int): EthUInt256 {
+        var value = EthUInt256.ZERO
+        var continuation: OwnershipContinuation? = null
+        do {
+            val filter = OwnershipFilterByItem(
```
Comment (yakischik) [inline nft/listener/src/main/kotlin/com/rarible/protocol/nft/listener/service/item/ItemOwnershipConsistencyService.kt:65]: Does this filter include sorting by ID?

File: nft/listener/src/main/kotlin/com/rarible/protocol/nft/listener/service/item/ItemOwnershipConsistencyJobHandler.kt [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            do {
+                val items = getItemsBatch(state.continuation)
+                logger.info("Got ${items.size} items to check")
+                if (items.isEmpty()) {
+                    state.continuation = null
+                    state.latestChecked = Instant.now()
+                    break
+                }
+
+                items.forEach { item -> checkAndFixItem(item) }
```
Comment (yakischik) [inline nft/listener/src/main/kotlin/com/rarible/protocol/nft/listener/service/item/ItemOwnershipConsistencyJobHandler.kt:52]: Maybe async in several (configurable) threads?

**Review by yakischik — APPROVED** [review]

File: nft/listener/src/main/kotlin/com/rarible/protocol/nft/listener/service/item/ItemOwnershipConsistencyService.kt (hunk 1 above)
Comment (augur) [reply nft/listener/src/main/kotlin/com/rarible/protocol/nft/listener/service/item/ItemOwnershipConsistencyService.kt:71]: Agree, fixed

File: nft/listener/src/main/kotlin/com/rarible/protocol/nft/listener/service/item/ItemOwnershipConsistencyService.kt (hunk 2 above)
Comment (augur) [reply nft/listener/src/main/kotlin/com/rarible/protocol/nft/listener/service/item/ItemOwnershipConsistencyService.kt:65]: Yes,

```
    private fun OwnershipFilter.Sort.toMongoSort() =
        when (this) {
            OwnershipFilter.Sort.LAST_UPDATE -> Sort.by(
                Sort.Order.desc(Ownership::date.name),
                Sort.Order.desc(Ownership::id.name)
            )
        }
```

File: nft/listener/src/main/kotlin/com/rarible/protocol/nft/listener/service/item/ItemOwnershipConsistencyJobHandler.kt (hunk 3 above)
Comment (augur) [reply nft/listener/src/main/kotlin/com/rarible/protocol/nft/listener/service/item/ItemOwnershipConsistencyJobHandler.kt:52]: I want this job not to be expensive, one thread should be enough for new items, I suppose.
(Lets change it if we see it is too slow in practice)

**Pull request merged**